backends/cuda/ceed-cuda-vector.c

```diff
@@ -153,6 +153,39 @@ static int CeedVectorSetArray_Cuda(const CeedVector vec,
   return 1;
 }
 
+//------------------------------------------------------------------------------
+// Vector Take Array
+//------------------------------------------------------------------------------
+static int CeedVectorTakeArray_Cuda(CeedVector vec, CeedMemType mtype,
+                                    CeedScalar **array) {
+  int ierr;
+  CeedVector_Cuda *impl;
+  ierr = CeedVectorGetData(vec, (void *)&impl); CeedChk(ierr);
+
+  switch(mtype) {
+  case CEED_MEM_HOST:
+    if (impl->memState == CEED_CUDA_DEVICE_SYNC) {
+      ierr = CeedSyncD2H_Cuda(vec); CeedChk(ierr);
+    }
+    (*array) = impl->h_array;
+    impl->h_array = NULL;
+    impl->h_array_allocated = NULL;
```

Comment:
Not a memory leak why?

Reply:
Should it be invalid (and documented+enforced in the interface) to Take an array that is owned by the Vector?

Reply:
I'm OK with the user taking ownership of (and management of) memory we allocated. Should I not be? Using the interface is declaring that you wish to take responsibility for the array. (probably should be added to the documentation)

I'm not sure where the leak would be if we assumed the user asked for responsibility and we transfer control over to them.

Reply:
How does a user free device memory that they didn't allocate? Direct calls to `cudaFree` or `hipFree` (or ...)?

Reply:
That is what I would expect. I expect this interface to be used when

1) the user called SetArray

Or

2) the user wants to take control of the array

1 would be the typical use. I could set it to be the only possible use pretty easily if we want to be more restrictive. It would not impact current usage.

I think we implicitly assume that the user knows what sort of device array is being used in the interface currently. If we are in the case where CEED_MEM_DEVICE could be allocated and freed differently on mixed device hardware we will only segfault at runtime if the user makes a mess. Otherwise the user should know how to free memory on their device.

Reply:
Sounds fair. Can you just make sure the documentation is clear about the responsibility in that case?